python/fbprophet/diagnostics.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+import numpy as np
+import pandas as pd
+from functools import reduce
+from fbprophet import Prophet
+
+
+def _copy(model):
+    """Copy Prophet object
+
+    Parameters
+    ----------
+    model: Prophet class object.
+
+    Returns
+    -------
+    Prophet class object with the same parameter with model variable
+    """
+
+    # TODO(@teramonagi) this function should be moved into Prophet class as a method.
+    result = Prophet(
+        growth=model.growth,
+        n_changepoints=model.n_changepoints,
+        yearly_seasonality=model.yearly_seasonality,
+        weekly_seasonality=model.weekly_seasonality,
+        holidays=model.holidays,
+        seasonality_prior_scale=model.seasonality_prior_scale,
+        changepoint_prior_scale=model.changepoint_prior_scale,
+        holidays_prior_scale=model.holidays_prior_scale,
+        mcmc_samples=model.mcmc_samples,
+        interval_width=model.interval_width,
+        uncertainty_samples=model.uncertainty_samples
+    )
+    return result
+
+
+def _cutoffs(df, horizon, k, period):
+    """Generate cutoff dates
+
+    Parameters
+    ----------
+    df: pd.DataFrame with historical data
+    horizon: pd.Timedelta.
+        Forecast horizon
+    k: Int number.
+        The number of forecasts point.
+    period: pd.Timedelta.
+        Simulated Forecast will be done at every this period.
+
+    Returns
+    -------
+    list of pd.Timestamp
+    """
+    # Allocate memory for result list
+    result = [None] * k
+    # Last cutoff is 'latest date in data - horizon' date
+    cutoff = df['ds'].max() - horizon
+    result[0] = cutoff
+
+    for i in range(1, k):
+        cutoff -= period
+        # If data does not exist in data range (cutoff, cutoff + horizon]
+        if not (((df['ds'] > cutoff) & (df['ds'] <= cutoff + horizon)).any()):
+            # Next cutoff point is 'closest date before cutoff in data - horizon'
+            closest_date = df[df['ds'] <= cutoff].max()['ds']
+            cutoff = closest_date - horizon
+        result[i] = cutoff
```

Comment:
This is a good place to handle the case where k * period is more than the length of the time series. Instead of pre-allocating `result`, just initialize it as `result = [cutoff]` and then here in L82:
```
if cutoff < df['ds'].min():
    logger.warning('Not enough data for requested number of cutoffs! Using {}.'.format(k))
    break
result.append(cutoff)
```